torch/csrc/api/src/data/samplers/distributed.cpp

```diff
@@ -0,0 +1,190 @@
+#include <torch/data/samplers/distributed.h>
+#include <torch/serialize/archive.h>
+#include <torch/types.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <random>
+#include <vector>
+
+namespace torch {
+namespace data {
+namespace samplers {
+
+DistributedSampler::DistributedSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : size_(size), num_replicas_(num_replicas), rank_(rank), epoch_(0) {
+  if (allow_duplicates) {
+    local_sample_count_ =
+        static_cast<size_t>(std::ceil(size_ * 1.0 / num_replicas_));
+  } else {
+    local_sample_count_ =
+        static_cast<size_t>(std::floor(size_ * 1.0 / num_replicas_));
+  }
+}
+
+void DistributedSampler::set_epoch(size_t epoch) {
+  epoch_ = epoch;
+}
+
+size_t DistributedSampler::local_sample_count() {
+  return local_sample_count_;
+}
+
+DistributedRandomSampler::DistributedRandomSampler(
+    size_t size,
+    size_t num_replicas,
+    size_t rank,
+    bool allow_duplicates)
+    : DistributedSampler(size, num_replicas, rank, allow_duplicates),
+      begin_index_(0),
+      end_index_(0),
+      sample_index_(0) {
+  populate_indices();
+  // shuffle first time.
+  reset(size_);
+}
+
+optional<std::vector<size_t>> DistributedRandomSampler::next(
+    size_t batch_size) {
+  if (sample_index_ == end_index_) {
+    return nullopt;
+  }
+
+  size_t end = sample_index_ + batch_size;
+  if (end > end_index_) {
+    end = end_index_;
+  }
+
+  auto iter = all_indices_.begin();
+  std::vector<size_t> res(iter + sample_index_, iter + end);
+  sample_index_ = end;
+  return res;
+}
+
+void DistributedRandomSampler::reset(optional<size_t> new_size) {
+  size_t size = new_size.value_or(size_);
+  if (size != size_) {
+    size_ = size;
+    populate_indices();
+  }
+  std::minstd_rand rand(epoch_);
+  std::shuffle(all_indices_.begin(), all_indices_.end(), rand);
```

Comment:
I feel like we could use `torch::randperm` with a custom generator in here, like we do in Python. That might be faster.

Reply:
I looked at that and not sure unless we really measure this, specially since we need to copy data out of the tensor or use tensor slicing operations.